feat: Publish RFC for SJB 2.0 release (#1718)
SAP/cloud-sdk: docs-java/guides/5.0-upgrade-steps.mdx

Starting with version 5.0.0 the SAP Cloud SDK requires a Java version of at least 17.
:::

:::info Minimum CAP version
If you are using the Cloud Application Programming model (CAP) to build your application, please note that the SAP Cloud SDK version 5 requires a minimum CAP/CDS version of at least `2.4.0`.
If you are on an older CAP stack, please follow the [CAP migration guide](https://cap.cloud.sap/docs/java/migration#one-to-two) first.
:::

## Before You Update

> Before you increase the SAP Cloud SDK version to 5, it is recommended to first update to the latest available version of
> the SAP Cloud SDK 4 and migrate any usage of deprecated API.

Reviewer: ```suggestion
Before you increase the SAP Cloud SDK version to 5, it is recommended to first update to the latest available version of the SAP Cloud SDK 4 and migrate any usage of deprecated API.
```

To find our BOMs, search for the following entries in your maven `dependencyManagement` section: `com.sap.cloud.sdk:sdk-bom` and `com.sap.cloud.sdk:sdk-modules-bom`.
Update them directly to version `5.0.0` or the [latest version](https://central.sonatype.com/artifact/com.sap.cloud.sdk/sdk-bom).

<Tabs groupId="bomType" defaultValue="springBoot" values={[ { label: 'Spring Boot', value: 'springBoot', }, { label: 'SAP Java Buildpack', value: 'sjb', }]}>

<TabItem value="sjb">

If you are using the SAP Java Buildpack and produce a `war` deployment (e.g. when you require JCo for your application),
please replace the `sdk-sjb-bom` with the `cf-tomee7-bom` and the `sdk-bom` instead:

```xml title="pom.xml"
<dependencyManagement>

> 	<dependencies>

Reviewer 2: (Minor)
Format

```suggestion
  <dependencies>
```

    <!-- Cloud SDK + SAP Java Buildpack BOMs -->
    <dependencies>
        <dependency>
            <groupId>com.sap.cloud.sjb.cf</groupId>
            <artifactId>cf-tomee7-bom</artifactId>
            <version>use-latest-version-here</version>
            <type>pom</type>
            <scope>import</scope>
        </dependency>
        <dependency>
            <groupId>com.sap.cloud.sdk</groupId>
            <artifactId>sdk-bom</artifactId>
            <version>use-latest-version-here</version>
            <type>pom</type>
            <scope>import</scope>
        </dependency>
        <!-- Work around errors in Buildpack BOM -->
        <dependency>
            <groupId>org.codehaus.woodstox</groupId>
            <artifactId>stax2-api</artifactId>
            <version>4.2.2</version>
            <scope>compile</scope>

>         </dependency>

Author: Only for SOAP, added a comment

    </dependencies>
  </dependencies>
</dependencyManagement>
```

Change the target runtime from `tomee7` to `tomcat`, and the buildpack from `sap_java_buildpack` to `sap_java_buildpack_jakarta`:

```diff title="manifest.yml"
  buildpacks:
-   - sap_java_buildpack
+   - sap_java_buildpack_jakarta
  env:
-   TARGET_RUNTIME: tomee7
+   TARGET_RUNTIME: tomcat

> ```

Reviewer 2: (Question)

SAP Machine needs to be set too, doesn't it?

Author: Yes, the guide mentions updating to Java 17 earlier

    <!-- Cloud SDK + SAP Java Buildpack BOMs -->
    <dependencies>
        <dependency>
            <groupId>com.sap.cloud.sjb.cf</groupId>
            <artifactId>cf-tomee7-bom</artifactId>
            <version>use-latest-version-here</version>
            <type>pom</type>
            <scope>import</scope>
        </dependency>
        <dependency>
            <groupId>com.sap.cloud.sdk</groupId>
            <artifactId>sdk-bom</artifactId>
            <version>use-latest-version-here</version>
            <type>pom</type>
            <scope>import</scope>
        </dependency>
        <!-- Work around errors in Buildpack BOM -->
        <dependency>
            <groupId>org.codehaus.woodstox</groupId>
            <artifactId>stax2-api</artifactId>
            <version>4.2.2</version>
            <scope>compile</scope>
        </dependency>
        <dependency>
            <groupId>com.sap.cloud.security.xsuaa</groupId>
            <artifactId>api</artifactId>
            <scope>compile</scope>
        </dependency>
        <dependency>
            <groupId>com.sap.cloud.security</groupId>
            <artifactId>java-security</artifactId>
            <scope>compile</scope>

>         </dependency>

Author: `java-security` was in the `dependencies` section of the `application/pom.xml` and not in the `dependencyManagement` of the main `pom.xml`. Therefore I'm not sure this works, please review

To find our BOMs, search for the following entries in your maven `dependencyManagement` section: `com.sap.cloud.sdk:sdk-bom` and `com.sap.cloud.sdk:sdk-modules-bom`.
Update them directly to version `5.0.0` or the [latest version](https://central.sonatype.com/artifact/com.sap.cloud.sdk/sdk-bom).

<Tabs groupId="bomType" defaultValue="springBoot" values={[ { label: 'Spring Boot', value: 'springBoot', }, { label: 'SAP Java Buildpack', value: 'sjb', }]}>

<TabItem value="sjb">

> If you are using the SAP Java Buildpack and produce a `war` deployment (e.g. when you require JCo for your application),
> please replace the `sdk-sjb-bom` with the `cf-tomee7-bom` (This might change) and the `sdk-bom` [like so](spring-boot-war-deployment#adjust-dependency-scopes) instead.

Reviewer: ```suggestion
If you are using the SAP Java Buildpack and produce a `war` deployment (e.g. when you require JCo for your application) please replace the `sdk-sjb-bom` with the `cf-tomcat-bom` and the `sdk-bom` as documented [here](/docs/java/guides/manage-dependencies#notes-on-the-sap-java-buildpack) instead.
```

Do we really need `cf-tomee7-bom`? If not, referencing `cf-tomcat-bom` is maybe simpler.

I wouldn't link to `spring-boot-war-deployment` because we didn't update that guide in ages, in particular we didn't update it for Spring Boot 3, no idea if it still works.

Author: - I kept cf-tomee7-bom because it is everywhere in the docs, waiting for a potential name change when SJB 2.2 is released
- I just wanted to link to the dependency management section to not duplicate the section, I can duplicate it if you prefer

Reviewer: Alright, but I would omit the "(This might change)" text. It leaves the reader puzzled what exactly might change, when, why etc. 

Instead, the reader (rightfully) expects that these docs will be updated if there are relevant changes. So we should omit the comment IMO.

  - `com.sap.cloud.sdk.datamodel:openapi-generator-cli`
    - Call the generator directly as a maven plugin:
    - `mvn com.sap.cloud.sdk.datamodel:openapi-generator-maven-plugin:5.0.0:generate`

> - Pregenerated Clients, [generate and use your own OData VDM](#generate-and-use-your-own-odata-virtual-data-model) instead.

Reviewer 3: ```suggestion
- Pregenerated Clients, [generate and use your own OData clients](#generate-and-use-your-own-odata-virtual-data-model) instead.
```

```

The file specified under _serviceNameMappingFile_ will be created and populated, if it doesn't exist yet.
If you want to create a VDM that is compatible with the OData VDM the SAP Cloud SDK provided in version 4, you can use this [_serviceNameMappings.properties_](/resources/SDKv4-ODatav2-serviceNameMappings.properties).

</TabItem>
</Tabs>

If you want to generate a VDM that is compatible with the On-Premise OData VDM the SAP Cloud SDK provided in version 4, take special care of the `packageName` and the contents of the _serviceNameMappingFile_ to reflect what you are expecting.

> ### Handle OData Virtual Data Model Instantiation

Reviewer 2: (off-topic)

_VDM / Virtual Data Model_ is a reserved term that should no longer be used for SDK 5.0 onwards.

Author: What should we use instead?

Reviewer 3: either the full name like this or "pre-generated clients"

  </table>
</details>

> ## Generate and Use Your Own OData Virtual Data Model

Reviewer 3: ```suggestion
## Generate and Use Your Own OData Clients
```

### Generate Your Own OData Virtual Data Model

The SAP Cloud SDK no longer provides pre-generated OData Virtual Data Models.

> We made this decision to reduce the maintenance effort and to allow you to generate your own VDM with the exact version of the OData service you are using.

Reviewer 3: ```suggestion
We made this decision to reduce the maintenance effort and to allow you to generate your own OData clients with the exact version of the OData service you are using.
```

That way there is no need to wait for the next SAP Cloud SDK release to get the latest changes of an OData service you are consuming.

To generate your own OData VDM, you can use the [OData Generator Maven Plugin](../features/odata/vdm-generator) directly.

> You can use the following configuration to generate a VDM that is compatible with the OData VDM the SAP Cloud SDK provided in version 4:

Reviewer 3: ```suggestion
You can use the following configuration to generate an OData client that is compatible with the OData VDM the SAP Cloud SDK provided in version 4:
```

```

The file specified under _serviceNameMappingFile_ will be created and populated, if it doesn't exist yet.

> If you want to create a VDM that is compatible with the OData VDM the SAP Cloud SDK provided in version 4, you can use this [_serviceNameMappings.properties_](/resources/SDKv4-ODatav4-serviceNameMappings.properties).

Reviewer 3: ```suggestion
If you want to create an OData client that is compatible with the OData VDM the SAP Cloud SDK provided in version 4, you can use this [_serviceNameMappings.properties_](/resources/SDKv4-ODatav4-serviceNameMappings.properties).
```

```

The file specified under _serviceNameMappingFile_ will be created and populated, if it doesn't exist yet.

> If you want to create a VDM that is compatible with the OData VDM the SAP Cloud SDK provided in version 4, you can use this [_serviceNameMappings.properties_](/resources/SDKv4-ODatav2-serviceNameMappings.properties).

Reviewer 3: ```suggestion
If you want to create an OData client that is compatible with the OData VDM the SAP Cloud SDK provided in version 4, you can use this [_serviceNameMappings.properties_](/resources/SDKv4-ODatav2-serviceNameMappings.properties).
```

</TabItem>
</Tabs>

> If you want to generate a VDM that is compatible with the On-Premise OData VDM the SAP Cloud SDK provided in version 4, take special care of the `packageName` and the contents of the _serviceNameMappingFile_ to reflect what you are expecting.

Reviewer 3: ```suggestion
If you want to generate an OData client that is compatible with the On-Premise OData VDM the SAP Cloud SDK provided in version 4, take special care of the `packageName` and the contents of the _serviceNameMappingFile_ to reflect what you are expecting.
```
